### DISCUSSION TITLE
JWT 토큰을 발행할 때, 어떤 점을 고려해야 할까?

### DISCUSSION THREAD

**Discussion by beatmeJY** [dailyge/dailyge-server#107] [Information]
## 🔗 연관 PR
- [[DLG-16] 구글 OAuth2 로그인을 구현한다.](https://github.com/dailyge/dailyge-server/pull/9)
- [[DLG-48] 만료된 AccessToken는 RefreshToken으로 토큰을 재발급한다.](https://github.com/dailyge/dailyge-server/pull/33)

<br/><br/>

## ❓ 작성 이유
JWT는 인증 및 권한 부여 과정에서 널리 사용되지만, 저장 위치 및 보안 조치에 따라 다양한 공격에 노출될 수 있기 때문에 이를 적절히 이용하고 방어할 수 있도록 학습 및 팀원들에게 공유하기 위해 작성하였습니다.

<br/><br/>

### JWT 사용 시 고려해야 할 사항

1. 웹 브라우저에는 어떻게 저장할지 (`쿠키`, `세션스토리지` 등)
2. XSS 공격과 CSRF 공격에 대해서 어떻게 대응할 지
3. 리프레시 토큰은 어디에 저장할지 (`서버` 또는 `별도 저장소`)
4. [jwt.io와 같은 사이트에서 정보를 쉽게 확인할 수 있는 문제](https://github.com/dailyge/dailyge-server/discussions/122)
5. [토큰을 도난당했을 때 어떻게 할지](https://github.com/dailyge/dailyge-server/discussions/114)

우선 이러한 다양한 문제 상황이 있을 것 같고, 각각의 고려사항에 대해 생각과 고민을 정리해보겠습니다.

4, 5번에 경우 내용이 많아 Discussions을 따로 링크를 걸어 두었습니다

<br/><br/>

## 1. JWT는 웹 브라우저 어디에 저장하는 것이 좋을까?

HTTP의 특성은 무 상태 프로토콜이기 때문에 각 요청은 독립적으로 처리되며, 서버는 이전 사용자의 요청을 기억하지 못 함.
따라서 JWT와 같은 사용자를 구분할 수 있는 중요한 정보를 어디에 저장해야 안전할지를 판단해야 함.

이 때 보관소로 사용될 곳은 주로 `localStorage`와 `cookie`를 선택할 수 있음.

### 장점
- 언제든지 localStrorage에 접근할 수 있어 접근성이 좋음.
- 일반적으로 5MB 정도의 데이터를 저장할 수 있음.
- 인증이 필요한 요청에만 명시적으로 데이터를 전송할 수 있음 (불필요한 요청에서 JWT가 전송되는 것을 방지)

### 단점
- XSS 공격에 취약.
- 공격자가 악의적인 스크립트를 사용자 브라우저에서 실행 시켜 localStorage에 있는 accessToken을 탈취 가능.

<br />

#### 장점
- 매 요청 시 자동으로 헤더에 포함.
- `HttpOnly`를 사용하면 JavaScript에서 접근할 수 없어 `XSS`의 공격으로부터 보호됨.
- `Secure`을 통해 HTTPS 요청만 허락하고, `SameSite` 속성을 이용해 동일한 사이트에서만 쿠키를 전송하도록 막으면 `CSRF` 공격을 방지할 수도 있음.

#### 단점
- 매 요청 시 자동으로 헤더에 포함되어 `CSRF` 공격에 취약.
- 공격자가 Request URL을 안다면 사용자가 악의적인 링크를 클릭하도록 유도하여 쿠키를 전송할 수 있음.
- 쿠키의 크기는 약 4KB의 데이터만 저장할 수 있음.

<br/><br/>

### XSS 공격과 대처방안
- XSS란 주로 악의적인 스크립트를 사용자가 실행하게 만들어 사용자의 토큰을 탈취하는 공격임.
- 쿠키에 경우 JS에서 접근이 불가능하도록 `HttpOnly` 조건을 주면, 해당 공격으로부터 안전함.
- localStorage의 경우 콘텐츠 보안 정책(CSP) 설정하던가, 입력 데이터를 검증하는 등의 별도의 방어 로직이 필요.

### CSRF 공격과 대처방안
- CSRF란 악의적인 웹사이트가 사용자의 브라우저에 이미 인증된 상태를 이용하여 원치 않는 요청을 보내게 만드는 공격임.
- localStorage의 경우 원하는 요청에만 토큰을 전송할 수 있어 CSRF의 공격으로부터 안전함
- 쿠키의 경우 매 요청마다 자동으로 전송되어 CSRF에 취약할 수 있음
- 따라서 쿠키의 경우 `Secure` 설정과 `SameSite` 설정 등을 통하여 CSRF공격에 대비해야 함.

<br/><br/>

## 3. RefreshToken은 어디에 저장할까?
* `RefreshToken`가 클라이언트에만 저장되어 있다면 해당 토큰이 유출될 경우 그만큼 오랫동안 피해를 입음.
* 따라서 `RefreshToken`은 서버에서 저장해두고 해당 값과 동일한 토큰이 전달받았는지를 확인해야함.

#### 장점
 - 서버에서 직접 토큰을 관리하기 때문에 접근과 성능에 용이.
 - 토큰에 접근할 수 있는 경로가 줄어들어 토큰이 유출될 가능성이 적어짐
 - 서버 내에서 관리하기에 만료된 토큰의 갱신 무효과 작업이 간편함
#### 단점
 - 서버에 직접 저장하는 경우는 보통 단일 서버이기 때문에 서버의 장애가 모든 서버의 장애로 이루질 수 있다.

<br />

#### 장점
 - 여러 서버나 서비스 간에 리프레시 토큰을 공유하기 쉬워 쉽게 서버 확장이 가능함.
 - 외부 저장소가 관리하는 백업, 복구, 확장성 등의 다양한 관리 기능을 활용할 수 있음.
#### 단점
 - 외부 저장소와 서버 간의 통신이 필요하므로 별도의 보안적인 인프라를 가지고 있어야 함.
 - 네트워크를 통해 접근해야 하므로 서버 내에 있는 메모리보다는 접근 시간이 길어질 수 있음.
 - 외부 저장소의 서비스가 장애 시 모든 서버의 장애로 이어질 수 있음.

<br /><br />

### 웹 브라우저에서의 저장 위치
 - XSS 공격에 안전한 **`Cookie`에 저장**
 - `HttpOnly`와 `Secure` 설정을 두어 안전하게 보관 및 사용함.

### 서버에서 저장 위치
 - 외부 저장소인 **레디스를 선택.**
 - 단일 서버의 경우 서버에 저장하는 방법도 효과적이지만, 웹 서비스의 경우 많은 사용자가 사용할 수 있다.
 - 따라서 리프레시 토큰은 서버의 확장을 고려해야 하기에 외부 저장소에 저장하는 것을 선택함.

**Comment by devjun10** [comment]
깔끔!
